src/main/java/org/apache/mesos/state/StateStoreCache.java

```diff
@@ -0,0 +1,301 @@
+            }
+        } finally {
+            RWLOCK.unlock();
+        }
+    }
+
+    @Override
+    public void storeStatus(TaskStatus status) throws StateStoreException {
+        RWLOCK.lock();
+        try {
+            store.storeStatus(status);
+            // corner case: only store status if a name=>id mapping already exists from storing taskinfo.
+            // this shouldn't come up in practice since the underlying statestore should have thrown anyway.
+            if (!idToTask.containsKey(status.getTaskId())) {
+                throw new StateStoreException(String.format(
+                        "The following TaskInfo is not present in the StateStore: %s. " +
+                        "TaskInfo must be present in order to store a TaskStatus.", status.getTaskId()));
+            }
+            idToStatus.put(status.getTaskId(), status);
+        } finally {
+            RWLOCK.unlock();
+        }
+    }
+
+    @Override
+    public void clearTask(String taskName) throws StateStoreException {
+        RWLOCK.lock();
+        try {
+            store.clearTask(taskName);
+            TaskID taskId = nameToId.remove(taskName);
+            if (taskId != null) {
+                idToTask.remove(taskId);
+                idToStatus.remove(taskId);
+            } else {
+                logger.warn("Unable to find task named {} to remove. Known task names are: {}",
+                        taskName, nameToId.keySet());
+            }
+        } finally {
+            RWLOCK.unlock();
+        }
+    }
+
+    @Override
+    public Collection<String> fetchTaskNames() throws StateStoreException {
+        RLOCK.lock();
+        try {
+            return nameToId.keySet();
+        } finally {
+            RLOCK.unlock();
+        }
+    }
+
+    @Override
+    public Collection<TaskInfo> fetchTasks() throws StateStoreException {
+        RLOCK.lock();
+        try {
+            return idToTask.values();
+        } finally {
+            RLOCK.unlock();
+        }
+    }
+
+    @Override
+    public Optional<TaskInfo> fetchTask(String taskName) throws StateStoreException {
+        RLOCK.lock();
+        try {
+            TaskID taskId = nameToId.get(taskName);
+            if (taskId == null) {
+                // Entry not found.
+                return Optional.empty();
+            }
+            TaskInfo task = idToTask.get(taskId);
+            if (task == null) {
+                // If we have a name->ID mapping, we really should have a TaskInfo for that ID.
+                throw new StateStoreException(String.format(
+                        "Cache consistency error: Unable to find TaskInfo for known ID '%s'. Name=>ID[%s] TaskIDs[%s]",
+                        taskName, nameToId, idToTask.keySet()));
+            }
+            return Optional.of(task);
+        } finally {
+            RLOCK.unlock();
+        }
+    }
+
+    @Override
+    public Set<TaskStatus> getTaskStatuses() throws Exception {
```

Comment:
Can we de-dup `getTaskStatuses` vs `fetchStatuses` ? Or there's a good reason why both should exist ?